[Sema] Group the IPI public-import diagnostic and emit it as a warning (#89648)
swiftlang/swift: userdocs/diagnostics/ipi-verification.md

# Public imports of project-internal modules (IPIVerification)

## Overview

> Modules built with `-library-level ipi` are project-internal, i.e. they are not intended to be redistributed, and their interfaces must not leak through the api/spi modules. Resolve this warning by hiding the dependency on the IPI module so it does not appear in the importing module's `.swiftinterface`.

Reviewer: ```suggestion
Modules built with `-library-level ipi` are considered project-internal and therefore are not intended to be redistributed. As a result, modules that _are_ intended for redistribution should not import `-library-level ipi` modules publicly since clients will not be able to resolve references to them.
```

> When an `api`- or `spi`-level module imports an `ipi`-level module in a way that exposes it to clients (a bare `public` import, an explicit `public import`, or an `@_exported import`), the compiler emits a warning. For example:

Reviewer: ```suggestion
When a module built with `-library-level api` or `-library-level api` imports a `-library-level ipi` module in a way that exposes the dependency to clients (a bare `public` import, an explicit `public import`, or an `@_exported import`), the compiler emits a warning. For example:
```

```swift
// MainLib is built with -library-level api
public import ProjectInternalUtilities
  // |- warning: Project internal module 'ProjectInternalUtilities' cannot be imported publicly because 'MainLib' has '-library-level api'
```

There are several ways to resolve this warning:

> If the dependency is purely an implementation detail and is never referenced in `public` or `@inlinable` API, mark the import `internal` (or rely on `InternalImportsByDefault`, where a bare `import` is already internal). The import will not appear in the generated `.swiftinterface`, so clients never see the IPI module:

Reviewer: I wouldn't refer to `.swiftinterface` files directly in this documentation, as they are a compilation implementation detail that developers aren't necessarily expected to be familiar with or interact with directly under most circumstances. I think diagnostic documentation should stick to language semantics and only delve into implementation details when its unavoidable.

```suggestion
If the dependency is purely an implementation detail and is never referenced in `public` or `@inlinable` API, mark the import `internal` (or adopt the upcoming feature `InternalImportsByDefault`, where a bare `import` is already `internal`).
```

```swift
internal import ProjectInternalUtilities
```

> If the dependency must be linked but should never be visible to clients in any form, use `@_implementationOnly import`. This is stronger than `internal` — types from the imported module cannot appear in any signature reachable from the public ABI:

Reviewer: Hm, I know we've been doing some work to make the functionality of `@_implementationOnly import` suitable for this purpose for some narrow use cases, but I wouldn't expect this recommendation to appear in public documentation. I would drop this section for now since for most projects `@_implementationOnly import` is still deprecated.

```swift
@_implementationOnly import ProjectInternalUtilities
```

If the importing module is itself not meant to be redistributed, lower its library level to `ipi`. Imports between non-distributable modules are unrestricted because the result is also non-distributable:

```sh
swiftc -library-level ipi ...
```

## Clang Modules

> Clang modules do not have their own `-library-level` setting. To mark a Clang module as project-internal, list it with `-ipi-clang-module <name>` when compiling the importer. Listed modules are then subject to the same check as Swift IPI modules:

Reviewer: ```suggestion
Clang modules do not have their own `-library-level` setting. To indicate to the compiler that a Clang module is not intended for redistribution, pass `-ipi-clang-module <module name>` when compiling the Swift modules that import it. Imports of the Clang module will be diagnosed the same way that `-library-level ipi` Swift modules are:
```

```sh
swiftc -library-level api -ipi-clang-module ProjectInternalCUtilities ...
```

```swift
// In a module built with the flags above:
public import ProjectInternalCUtilities
  // |- warning: Project internal module 'ProjectInternalCUtilities' cannot be imported publicly because 'MainLib' has '-library-level api'
```

> The same resolutions apply: mark the import `internal` or `@_implementationOnly`, or lower the importer's library level. Clang modules not listed via `-ipi-clang-module` are unaffected by this check.

Reviewer: I don't think this needs to be re-iterated. Maybe the section on resolving the warning could go below this instead?

Reviewer: This documentation could use a "See Also" section with links to documentation/Swift evolution proposals for `InternalImportsByDefault` and access levels on imports.

Author: I'll also add a link to IPI library level proposal when/if it is published
